### ISSUE TITLE
GUI ?

### ISSUE THREAD

**Issue by baykus871** [h3rald/min#70]
I am tiring you with my questions, but somethings can be made in this way:

How can i do GUI doing? or "making?"(but don't forget i am a little good at Nim)

I feeling this, GUI, is a component which really hard to make. Am i right about my concerns?

**Comment by h3rald** [comment]
Believe me, I did think about it. I would love to have a **minimal** UI toolkit to be included in min!

I am currently looking into potential candidates, but it is quite hard to find something that is:

* sufficiently small, I don't want the size of the min executable to grow too much (OpenSSL was bad enough)
* Cross platform, it must run at least on x64 Linux, Windows, and Mac OS X
* Ideally have no dependencies and/or leverage existing native platform widgets 

I was looking at something like [mui](https://github.com/angluca/mui), for example... basically it would be a matter of exposing the existing Nim API in min.

I also considered going the picolisp route: picolisp comes with a simple library that makes you write html code using native (lisp) syntax: https://picolisp.com/wiki/?guiscripting

The idea is then to serve it from a local web server running on localhost. As a matter of fact you could already build something like that with min now (I am experimenting building an API server, it seems to work *enough*). The only problem I have with it is... well, I have some security concerns. Running a web server locally with potentially unlimited access to the file system kinda scares me, but technically unless it is exposed in some way to the outside world (e.g. a reverse proxy) or someone gets access to localhost in some way it shouldn't be too bad.

What were you looking for? What kind of capabilities?

**Comment by baykus871** [comment]
I am stranger to GUI however i can learn how. My plan is making an IDE to Min. If happens, then Min coding will be easier more.

**Comment by h3rald** [comment]
Oh well... that's a very challenging task for sure!

I built my own web server, data store, programming language, static site generator, documentation tool, package manager but... never an IDE or an editor! Much easier to create extensions for the existing ones! Did you try the min syntax highlighting file for vim? And also I created an extension for Visual Studio Code as well:

https://marketplace.visualstudio.com/items?itemName=h3rald.vscode-min-lang

**Comment by baykus871** [comment]
Can you do debugging Min in VSCode ? (I think not, but must be)

If we a knowing about stacks then it will be **easier**.

**Comment by h3rald** [comment]
No, but it would be possible to create a sort of min server to do things like auto-complete, suggestions, interpreting code on-the-fly and warn users... someday. It is possible for Nim and all other languages, so it could be done with min as well. For now I only focused on highlighting though...
